tests/zsa.rs

```diff
@@ -0,0 +1,700 @@
+    };
+
+    // Verify the shielded bundle
+    verify_bundle(&shielded_bundle, &vk);
+}
+
+/// Issue single ZSA note and split it into 2 notes
+#[test]
+fn e2e_1zsa_to_2zsa() {
+    let mut rng = OsRng;
+    let (pk, vk, sk, fvk, recipient, isk, ik) = prepare_keys();
+    let asset_descr = "zsa_asset";
+
+    // Create a issuance bundle
+    let mut unauthorized = IssueBundle::new(ik);
+
+    // Add output ZSA note
+    assert!(unauthorized
+        .add_recipient(
+            asset_descr.to_string(),
+            recipient,
+            NoteValue::from_raw(42),
+            false,
+            &mut rng,
+        )
+        .is_ok());
+
+    let issuing_bundle = sign_issuing_bundle(unauthorized, rng, isk);
+
+    // Take first note from first action
+    let note = issuing_bundle.actions().first().unwrap().notes().first();
+
+    let (merkle_path, anchor) = build_merkle_tree(note);
+
+    verify_issuing_bundle(&issuing_bundle);
+
+    // Create a shielded bundle spending the previous output
+    let shielded_bundle: Bundle<_, i64> = {
+        let mut builder = Builder::new(Flags::from_parts(true, true), anchor);
+        // Add single ZSA spend
+        assert_eq!(builder.add_spend(fvk, *note, merkle_path), Ok(()));
+        // and 2 ZSA outputs
+        assert_eq!(
+            builder.add_recipient(
+                None,
+                recipient,
+                NoteValue::from_raw(40),
+                note.note_type(),
+                None
+            ),
+            Ok(())
+        );
+        assert_eq!(
+            builder.add_recipient(
+                None,
+                recipient,
+                NoteValue::from_raw(2),
+                note.note_type(),
+                None
+            ),
+            Ok(())
+        );
+        sign_shielded_bundle(builder, rng, pk, sk)
+    };
+
+    // Verify the shielded bundle
+    verify_bundle(&shielded_bundle, &vk);
+}
+
+/// Issue 2 ZSA notes and join them into a single note
+#[test]
+fn e2e_2zsa_to_1zsa() {
+    let mut rng = OsRng;
+    let (pk, vk, sk, fvk, recipient, isk, ik) = prepare_keys();
+    let asset_descr = "zsa_asset";
+
+    // Create a issuance bundle
+    let mut unauthorized = IssueBundle::new(ik);
+
+    // Add 2 output ZSA notes
+    assert!(unauthorized
+        .add_recipient(
+            asset_descr.to_string(),
+            recipient,
+            NoteValue::from_raw(40),
+            false,
+            &mut rng,
+        )
+        .is_ok());
+    assert!(unauthorized
+        .add_recipient(
+            asset_descr.to_string(),
+            recipient,
+            NoteValue::from_raw(2),
+            false,
+            &mut rng,
+        )
+        .is_ok());
+
+    let issuing_bundle = sign_issuing_bundle(unauthorized, rng, isk);
+
+    // Take notes from first action
+    let note1 = issuing_bundle
+        .actions()
+        .first()
+        .unwrap()
+        .notes()
+        .get(0)
+        .unwrap();
+    let note2 = issuing_bundle
```

Comment:
Please create a getter to return all notes from all actions in `issue_bundle`.
`let notes :: Vec<&Notes> = issuing_bundle.get_all_notes();`
and replace in all positions.

We will also need this down the road for consensus tree building.

Reply:
done